django/db/models/sql/query.py

```diff
@@ -1299,13 +1299,32 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
             self.check_filterable(value)
 
         clause = self.where_class()
+        alias = self.get_initial_alias()
         if reffed_expression:
             condition = self.build_lookup(lookups, reffed_expression, value)
             clause.add(condition, AND)
-            return clause, []
+            targets = reffed_expression.get_source_expressions()
+            if targets and isinstance(targets[0], Col):
+                target = targets[0].target
+                lookup_type = condition.lookup_name
+                clause.add(condition, AND)
+                if current_negated and lookup_type != 'isnull' and condition.rhs is not None \
+                        and self.is_nullable(target):
+                    # The condition added here will be SQL like this:
+                    # NOT (col IS NOT NULL), where the first NOT is added in
+                    # upper layers of code. The reason for addition is that if col
+                    # is null, then col != someval will result in SQL "unknown"
+                    # which isn't the same as in Python. The Python None handling
+                    # is wanted, and it can be gotten by
+                    # (col IS NULL OR col != someval)
+                    #   <=>
+                    # NOT (col IS NOT NULL AND col = someval).
+                    lookup_class = target.get_lookup('isnull')
+                    col = self._get_col(target, target, alias)
+                    clause.add(lookup_class(col, False), AND)
```

Comment:
Have you explored a way to resolve the DRY issues here?

Reply:
Originally i try to use the below logic for checking negated null column.
```
        if current_negated and (lookup_type != 'isnull' or condition.rhs is False) and condition.rhs is not None:
            require_outer = True
            if lookup_type != 'isnull':
                # The condition added here will be SQL like this:
                # NOT (col IS NOT NULL), where the first NOT is added in
                # upper layers of code. The reason for addition is that if col
                # is null, then col != someval will result in SQL "unknown"
                # which isn't the same as in Python. The Python None handling
                # is wanted, and it can be gotten by
                # (col IS NULL OR col != someval)
                #   <=>
                # NOT (col IS NOT NULL AND col = someval).
                if (
                    self.is_nullable(targets[0]) or
                    self.alias_map[join_list[-1]].join_type == LOUTER
                ):
                    lookup_class = targets[0].get_lookup('isnull')
                    col = self._get_col(targets[0], join_info.targets[0], alias)
                    clause.add(lookup_class(col, False), AND)
                # If someval is a nullable column, someval IS NOT NULL is
                # added.
                if isinstance(value, Col) and self.is_nullable(value.target):
                    lookup_class = value.target.get_lookup('isnull')
                    clause.add(lookup_class(value, False), AND)
```

but in my view, other logic on below that code is not designed for reffed_expression.
so, I can't resolve DRY issues.
If you have any good idea, please tell me.
```
        try:
            join_info = self.setup_joins(
                parts, opts, alias, can_reuse=can_reuse, allow_many=allow_many,
                reuse_with_filtered_relation=reuse_with_filtered_relation,
            )

            # Prevent iterator from being consumed by check_related_objects()
            if isinstance(value, Iterator):
                value = list(value)
            self.check_related_objects(join_info.final_field, value, join_info.opts)

            # split_exclude() needs to know which joins were generated for the
            # lookup parts
            self._lookup_joins = join_info.joins
        except MultiJoin as e:
            return self.split_exclude(filter_expr, can_reuse, e.names_with_path)
```